test/common_distributed.py

```diff
@@ -125,31 +125,52 @@ def setUpClass(cls):
                 fn = getattr(cls, attr)
                 setattr(cls, attr, cls.join_or_run(fn))
 
+    def __init__(self, *args, **kwargs):
+        super(MultiProcessTestCase, self).__init__(*args, **kwargs)
+
+        self.rank = self.MAIN_PROCESS_RANK
+
     def setUp(self):
         super(MultiProcessTestCase, self).setUp()
         self.skip_return_code_checks = []
-        self.rank = self.MAIN_PROCESS_RANK
-        self.file = tempfile.NamedTemporaryFile(delete=False)
-        self.processes = [self._spawn_process(rank) for rank in range(int(self.world_size))]
+        if self.rank == self.MAIN_PROCESS_RANK:
+            # spawn can not pass file handle, so pass file name here
+            self.file_name = tempfile.NamedTemporaryFile(delete=False).name
+            self.processes = [self._spawn_process(rank) for rank in range(int(self.world_size))]
 
     def tearDown(self):
         super(MultiProcessTestCase, self).tearDown()
         for p in self.processes:
             p.terminate()
 
+    def _current_test_name(self):
+        # self.id() == e.g. '__main__.TestDistributed.TestAdditive.test_get_rank'
+        return self.id().split(".")[-1]
+
     def _spawn_process(self, rank):
         name = 'process ' + str(rank)
-        process = multiprocessing.Process(target=self._run, name=name, args=(rank,))
+        test_name = self._current_test_name()
+        process = multiprocessing.Process(
+            target=self.__class__._run,
+            name=name,
+            args=(test_name, rank, self.file_name)
+        )
         process.start()
         return process
 
-    def _run(self, rank):
+    @classmethod
+    def _run(cls, test_name, rank, file_name):
+        self = cls(test_name)
+
         self.rank = rank
+        self.file_name = file_name
+
+        self.setUp()
 
-        # self.id() == e.g. '__main__.TestDistributed.test_get_rank'
-        # We're retreiving a corresponding test and executing it.
-        getattr(self, self.id().split(".")[2])()
-        sys.exit(0)
+        # We're retrieving a corresponding test and executing it.
+        getattr(self, test_name)()
+        if multiprocessing.get_start_method() != "spawn":
+                sys.exit(0)
```

Comment:
Doesn't this affect all c10d tests? Also, why are we checking for this after the test is done instead of before it?

Reply:
c10d is not spawn, so it is not affected. we just change behavior for rpc, when it is spawn, we do not need to exit() here explicitly, as @xush6528 mentioned sys.exit(0) here is mainly for fork processes to avoid run teardown().